cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,192 @@
+ * their application. This is more efficient than opening and closing instances for the same
+ * underlying file system location multiple times throughout the lifetime of the application.
+ * Users should call {@link #close()} after using a store to shut down in an orderly manner and free
+ * up resources.
+ * </p>
+ *
+ * @api_public
+ */
+public class DocumentStore {
+
+    public final Database database;
+    public final IndexManager query;
+    protected final String databaseName; // only used for events
+    private final File location; // needed for close
+
+    /* This map should only be accessed inside synchronized(openDatastores) blocks */
+    private static final Map<File, DocumentStore> documentStores = new HashMap<File, DocumentStore>();
+
+    private static EventBus eventBus = new EventBus();
+
+    private static final Logger logger = Logger.getLogger(DocumentStore.class.getCanonicalName());
+
+    private DocumentStore(File location, KeyProvider keyProvider) throws DatastoreException, IOException, SQLException {
+        this.location = location;
+        this.databaseName = location.getName();
+        this.database = new DatabaseImpl(location, keyProvider);
+        this.query = new IndexManagerImpl(database);
+    }
+
+    /**
+     * <p>
+     * Get an instance of an existing or newly created store.
+     * </p>
+     * <p>
+     * Equivalent to calling {@link #getInstance(File, KeyProvider)} getInstance} with a
+     * {@link NullKeyProvider}.
+     * </p>
+     * @param location The location on the file system where the underlying files should be stored.
+     *                 Must be a directory.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location) throws DatastoreNotCreatedException {
+        return getInstance(location, new NullKeyProvider());
+    }
+
+    /**
+     * <p>
+     * Get an instance of an existing or newly created store.
+     * </p>
+     * <p>
+     * If encryption is enabled for this platform, passing a {@link KeyProvider} containing a non-null
+     * key will open or create an encrypted database. If opening a database, the key used to
+     * create the database must be used. If encryption is not enabled for this platform, returning
+     * a non-null key will result in an exception.
+     * </p>
+     * @param location The location on the file system where the underlying files should be stored.
+     *                 Must be a directory.
+     * @param provider KeyProvider object. Use a {@link NullKeyProvider} if the database shouldn't
+     *                 be encrypted.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
```

Comment:
DatastoreNotCreatedException does not seem like the correct type for cases where it exists, but can't be opened for some reason.